### DISCUSSION TITLE
stacking cross correlations

### DISCUSSION THREAD

**Discussion by vbrossard** [PeterMakus/SeisMIC#74] [Q&A] [answered]
Hello everyone,

I am currently working with SeisMIC and have a question regarding the stacking function that I hope you can help clarify. I want to plot the stacking of the cross-correlation for a month, but the date shown in the figure suggests that the code has performed the stacking for the entire year over which I have computed the cross-correlation. The .h5 file contains cross-correlation data for one year.

This is part of the code I am using for plotting the stacking

###PLOTTING
for net in networks :
    for sta in stations :
        for com in comps :
            #name of the correlation file
            corr_name = f"{net}.{sta}.{com}.h5" 

            #reading the correlation functions
            with CorrelationDataBase(f'/home/vi/SeisMIC-main/{corr_dir}/{corr_name}', mode='r') as cdb:
                print(list(cdb.keys()))
                 
                cst = cdb.get_data(tag='subdivision', network=net, station=sta, channel=com, corr_start="2020-01-01T00:00:00",       corr_end="2020-01-31T23:59:59")
                #cst is a CorrStream object on that we can use our known methods
                 
            plt.figure()
            cst.plot(timelimits=[- time_limits, time_limits], cmap='seismic')
            plt.title(f'correlation {net}.{sta}.{com}')
            plt.savefig(f'/home/vi/SeisMIC-main/{corr_dir}/corr.{net}.{sta}.{com}.png')
           
            
            plt.figure()
            stime = UTCDateTime("2020-01-01T00:00:00")
            print(stime)
            etime = UTCDateTime("2020-01-31T23:59:59")
            print(etime)
            corrstack = cst.stack(starttime = stime, endtime = etime)[0]
            corrstack.plot(tlim=[-250, 250])
            plt.savefig(f'/home/vi/SeisMIC-main/{corr_dir}/corr.{net}.{sta}.{com}.stack.png')

I appreciate any insights or resources you could share on these topics. Thank you.

**Comment by PeterMakus** [comment, marked as answer]
Hi there,
so here is what happens: the kwargs ``corr_start``and ``corr_end``in ``cdb.get_data``are expected to be of type ``UTCDateTime``. If they are not, they will be replaced by wildcards -> all available data will be loaded into the stream. I admit that this is a little unfortunate and, in a future version, I will let the code print a warning, when this happens.

However, even when adding them as UTCDateTime objects, you will still run into problems. ``corr_start`` and ``corr_end``refer to the (almost) exact start and endtime of one CorrTrace. Currently, there is no way of filtering for all traces in a certain time span - I acknowledge that this is inconvenient but this is actually a surprisingly complex problem.

Here are two ways for a workaround:
Probably the easiest and most stable would be load in all data and then select the times you want to keep after. This will take a little longer because you load unnecessary data but the extra time is not particularly significant for all but the largest datasets. You can do that like so:

1. Load in data like you did using ``db.get_data``but insert wildcards ('*') for corr_start and corr_end
2. Filter the resulting Corrstream using it's method ``cst.select_corr_time``(see [docs](https://petermakus.github.io/SeisMIC/modules/API/API.html#seismic.correlate.stream.CorrStream.select_corr_time).

The slightly more complex but somewhat faster way would be to find out what times are available in you file and then loop over the list of Traces that you would like to load.
1. Use ``cdb.get_available_starttimes` [see](https://petermakus.github.io/SeisMIC/modules/API/API.html#seismic.db.corr_hdf5.DBHandler.get_available_starttimes). This returns a dictionary
2. Use the output of the dictionary to loop over get_data. Make sure to not open and close the hdf file after each iteration as this will lead to a lot of unnecessary overhead.

I would recommend going for the first solution.
I hope that clarifies the situation a little.

Peter

**Reply by vbrossard** [reply]
Thank you very much for your quick response. I wanted to let you know that your first suggestion worked perfectly, and it resolved the issue I was facing. I really appreciate your help and the time you took to address my question so promptly.
